### PR TITLE
Jgfouca/merge attempt

### PR DESCRIPTION
**Pull request by jgfouca**
Bring in changes from ACME/cime between 1/9/16 and 3/23/16.

coupler: fix an error message, modify placement of t_adj calls in the driver, turn on driver threading by default.
timer: add option to prefix timer names with detail level, Update progress monitoring of log files.
grids: add mpas-ocean grid and mapping file settings for oQU240. ACME low-res and mapping files.
compsets: Add ACME coupled A_B* compsets, MOSART compsets, WCYCL, CRYO compsets.
tests: Update atm_developer test suite and compsets including 72L compsets. Add atm tests.
scripts: Changes to pointCLM.py and site_fullrun.py, add RESUBMIT_QUEUED, work with HOMME repo.
utils: bug fix in cprnc
machines: Add OpenACC machines files to Titan
Updates to various machines files, edison, cori, mira, titan, redsky, skybridge, constance.
Lots of work on edison, cori modules.
nested OMP flags.
remove -V from PBS

### CODE REVIEW

File: cime_config/cesm/machines/config_batch.xml [hunk 1]
```diff
@@ -91,6 +91,7 @@
       <directive> -j oe </directive>
       <directive default="ae"  > -m {{ mail_options }} </directive>
       <directive> -V </directive>
+      <directive default="/bin/bash" > -S {{ shell }} </directive>
```
Comment (jgfouca) [inline cime_config/cesm/machines/config_batch.xml]: Not sure how this got added by this merge. Should I remove?

File: driver_cpl/driver/cesm_comp_mod.F90 [hunk 2]
```diff
@@ -776,7 +776,7 @@ subroutine cesm_pre_init2()
         nthreads_LNDID,nthreads_ICEID,nthreads_OCNID,nthreads_GLCID, &
         nthreads_ROFID, nthreads_WAVID, pethreads_GLOID )
 
-   call t_initf(NLFileName, LogPrint=.false., mpicom=mpicom_GLOID, &
```
Comment (jgfouca) [inline driver_cpl/driver/cesm_comp_mod.F90]: Please look at this file closely.

File: scripts/Testing/Testcases/HOMME_build.csh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/csh -f
+
+./Tools/check_lockedfiles || exit -1
+
+setenv COMPILER `./xmlquery COMPILER -value`
+source env_mach_specific || exit -1
+
```
Comment (jgfouca) [inline scripts/Testing/Testcases/HOMME_build.csh]: Needs to be ported to python at some point, not urgent.

File: scripts/Tools/case.submit [hunk 4]
```diff
@@ -115,7 +116,12 @@ sub main
 	    check_DA_settings();
 	    `./xmlchange --noecho  RUN_WITH_SUBMIT=TRUE`;
 	    $config{RUN_WITH_SUBMIT}='TRUE';
-	    $batchutils->submitJobs();
+            if (defined $opts{'jobid'}) {
```
Comment (jgfouca) [inline scripts/Tools/case.submit]: Will this break stuff?

File: scripts/Tools/pointclm/pointCLM.py [hunk 5]
```diff
@@ -132,36 +132,23 @@
 #                   default=False,action="store_true")
 parser.add_option("--nofire", action="store_true", dest="nofire", default=False, \
                     help="To turn off wildfires")
-parser.add_option("--npoolmod", action="store_true", dest="npoolmod", default=False, \
-                    help="To turn on nitrogen pool modifications")
-parser.add_option("--cpoolmod", action="store_true", dest="cpoolmod", default=False, \
-                    help="To turn on carbon storage pool modifications")
-
-parser.add_option("--q10wbmod", action="store_true", dest="q10wbmod", default=False, \
```
Comment (jgfouca) [inline scripts/Tools/pointclm/pointCLM.py]: Changes to this file are not relevant to the rest of CIME.

File: scripts/Tools/pointclm/site_fullrun.py [hunk 6]
```diff
@@ -59,26 +59,18 @@
                   help = 'Use C14 as C13 (no decay)', action="store_true")
 parser.add_option("--ninst", dest="ninst", default=1, \
                       help = 'number of land model instances')
-parser.add_option("--npoolmod", action="store_true", dest="npoolmod", default=False, \
-                    help="To turn on nitrogen pool modifications")
```
Comment (jgfouca) [inline scripts/Tools/pointclm/site_fullrun.py]: Ditto above.

File: share/timing/perf_mod.F90 [hunk 7]
```diff
@@ -135,6 +135,13 @@ module perf_mod
    logical, private   :: perf_ovhd_measurement = def_perf_ovhd_measurement
                          ! measure overhead of profiling directly
 
+   logical, parameter :: def_perf_add_detail = .false.         ! default
```
Comment (jgfouca) [inline share/timing/perf_mod.F90]: Needs careful review.

File: tools/cprnc/compare_vars_mod.F90.in [hunk 8]
```diff
@@ -119,7 +119,7 @@ contains
           if(.not. ignoretime) then
              do while(t1<=ns1 .and. t2<= ns2)
                 tdiff = abs(time(t1,1) - time(t2,2))
-                if ( tdiff <= timeepsilon) exit
+                if (tdiff < timeepsilon) exit
```
Comment (jgfouca) [inline tools/cprnc/compare_vars_mod.F90.in]: This was labelled as a bugfix.

**Comment by jgfouca** [comment]
Annotations complete. This passed all tests on melvin. It passed cime_developer on skybridge (batch machine).

File: utils/perl5lib/Module/ModuleLoader.pm [hunk 9]
```diff
@@ -655,15 +655,26 @@ sub loadSoftModules()
 
     my %newbuildenv;
 
+	# ndk: add a loop to see if our above module adjusting _removed_
```
Comment (jgfouca) [inline utils/perl5lib/Module/ModuleLoader.pm]: @ndkeen , why was this needed?

Comment (ndkeen) [inline utils/perl5lib/Module/ModuleLoader.pm]: After module commands (removing/loading), the script looks at the env variables.  Before and After.
Any new ones, it adds to the list.  Any changes to existing vars, it changes.
Any env vars that are _removed_ it would not previously remove them, which caused several issues.
This fix removes them.  Surely, new cime does not do things in this way.

Comment (jgfouca) [inline utils/perl5lib/Module/ModuleLoader.pm]: New CIME has a python module loader but I think the perl one is still used in code that hasn't been ported yet. It shouldn't be used much longer, so maybe we shouldn't worry.

File: cime_config/cesm/machines/config_batch.xml (hunk 1 above)
Comment (jedwards4b) [inline cime_config/cesm/machines/config_batch.xml]: Yes, since it's on the cesm side.

File: driver_cpl/driver/cesm_comp_mod.F90 (hunk 2 above)
Comment (jedwards4b) [inline driver_cpl/driver/cesm_comp_mod.F90]: Adjusting details of timer output, I think it's okay.

File: driver_cpl/shr/seq_flds_mod.F90 [hunk 10]
```diff
@@ -1335,6 +1335,15 @@ subroutine seq_flds_set(nmlfile, ID)
      attname  = 'Si_ifrac'
      call metadata_set(attname, longname, stdname, units)
 
+     ! Sea ice basal pressure
+     call seq_flds_add(i2x_states,"Si_bpress")
```
Comment (jedwards4b) [inline driver_cpl/shr/seq_flds_mod.F90]: This may need to be in an ifdef if it is not compatible with our cice component @mvertens what do you think?

File: externals/CMake/FindNETCDF.cmake [hunk 11]
```diff
@@ -19,7 +19,7 @@ find_library(NetcdfF_LIBRARY
 find_path(Netcdf_NC_CONFIG_BIN
           NAMES nc-config
           HINTS ${Netcdf_INCLUDE_DIR}/../bin
-          NO_SYSTEM_ENVIRONMENT_PATH NO_CMAKE_SYSTEM_PATH)
+          NO_CMAKE_SYSTEM_PATH)
```
Comment (jedwards4b) [inline externals/CMake/FindNETCDF.cmake]: This could lead to errors if a user has NETCDF_PATH set incorrectly in the environment.  Currently we are requiring it to be explicitly set in config_machines.xml or config_compilers.xml.

File: scripts/Testing/Testcases/HOMME_build.csh (hunk 3 above)
Comment (jedwards4b) [inline scripts/Testing/Testcases/HOMME_build.csh]: This test is a little outside the scope of system testing - it's more like a unit test for the homme dycore.   I don't think that it would work on the cesm side.   Perhaps it would be more appropriate in the unit testing framework.

File: scripts/Tools/case.submit (hunk 4 above)
Comment (jedwards4b) [inline scripts/Tools/case.submit]: I'm about to remove this script.  I think maybe someone was trying to improve the flawed logic here.   Is there a log note or an issue explaining what this was to address?

File: scripts/create_newcase [hunk 12]
```diff
@@ -82,10 +82,14 @@ def parse_command_line(args, cimeroot):
                         "with the new user grid changes added to it"
 			"This argument is required if --user-grid is True")
 
-    parser.add_argument("--srcroot",default=os.path.dirname(cimeroot),
+    parser.add_argument("--srcroot", default=os.path.dirname(cimeroot),
                         help="Alternative path for source root directory. By default this is set to"
                         "cimeroot/../")
 
+    if cime_model == "acme":
```
Comment (jedwards4b) [inline scripts/create_newcase]: This is a departure from the normal specification of model options.  If we go down this road we may open the door to any number of component specific features as options to create_newcase.   I think it would be better to make this a cam config option and a compset specification.

File: share/timing/perf_mod.F90 [hunk 13]
```diff
@@ -637,18 +657,38 @@ subroutine t_startf(event, handle)
 !---------------------------Local workspace-----------------------------
 !
    integer  ierr                          ! GPTL error return
+   integer  str_length, i                 ! support for adding
+                                          !  detail prefix
+   character(len=2) cdetail               ! char variable for detail
+   character(len=SHR_KIND_CX+4) aug_event ! augmented label
```
Comment (jedwards4b) [inline share/timing/perf_mod.F90]: don't like the +4 here

File: share/timing/perf_mod.F90 [hunk 14]
```diff
@@ -637,18 +657,38 @@ subroutine t_startf(event, handle)
...
 !
 !-----------------------------------------------------------------------
 !
    if (.not. timing_initialized) return
    if (timing_disable_depth > 0) return
 
+   if ((perf_add_detail) .AND. (cur_timing_detail < 100)) then
+
+      do i=1,SHR_KIND_CX+4
```
Comment (jedwards4b) [inline share/timing/perf_mod.F90]: yuch

File: utils/perl5lib/Module/ModuleLoader.pm (hunk 9 above)
Comment (jedwards4b) [inline utils/perl5lib/Module/ModuleLoader.pm]: We are just about done with the perl module loader.  We rely much much less on environment variables now so I hope we no longer need this kind of hack.

File: utils/python/CIME/case.py [hunk 15]
```diff
@@ -300,7 +300,7 @@ def _get_component_config_data(self):
     def configure(self, compset_name, grid_name, machine_name=None,
                   project=None, pecount=None, compiler=None, mpilib=None,
                   user_compset=False, pesfile=None,
-                  user_grid=False, gridfile=None):
+                  user_grid=False, gridfile=None, cam_se_target=None):
```
Comment (jedwards4b) [inline utils/python/CIME/case.py]: cam_se_target is inappropriate here.

File: utils/python/CIME/case.py [hunk 16]
```diff
@@ -357,6 +362,10 @@ def configure(self, compset_name, grid_name, machine_name=None,
         else:
             expect(machobj.is_valid_compiler(compiler),
                    "compiler %s is not supported on machine %s" %(compiler, machine_name))
+
+        if get_model() == "acme" and cam_se_target == "preqx_acc" and machine_name == "titan":
```
Comment (jedwards4b) [inline utils/python/CIME/case.py]: This should be in config_compilers.xml not in code.

File: utils/python/update_acme_tests.py [hunk 17]
```diff
@@ -73,13 +83,16 @@
                          "ERS_IOP4p.f19_g16_rx1.A",
                          "ERS_IOP4p.ne30_g16_rx1.A",
                          ("ERP_Ln9.ne30_ne30.FC5", "cam-outfrq9s"),
+                         "HOMME_P24.f19_g16_rx1.A",
```
Comment (jedwards4b) [inline utils/python/update_acme_tests.py]: won't work with CIME_MODEL=cesm

**Comment by jedwards4b** [comment]
I wish it were more collaborative in nature but I'm not too worried about the timing changes.   I am a little  more concerned about the homme test, which should be a unit test not a system test.   The cam_se_target really needs to be removed and implemented in the appropriate manner.

File: driver_cpl/shr/seq_flds_mod.F90 (hunk 10 above)
Comment (mvertens) [inline driver_cpl/shr/seq_flds_mod.F90]: It should not need an ifdef. The sea ice components are free not to fill in this field - and the ocean is free not to read it. So the only thing is a bit of computational expense for redistribution (since there is no regridding). I think a comment in the code specifying that this only applies to acme would be helpful.

File: scripts/Tools/pointclm/pointCLM.py (hunk 5 above)
Comment (mvertens) [inline scripts/Tools/pointclm/pointCLM.py]: This is used strictly by clm on our side. Why has this changed?

File: scripts/create_newcase (hunk 12 above)
Comment (mvertens) [inline scripts/create_newcase]: I very much agree with Jim.

**Comment by worleyph** [comment]
FYI - the "yuchy" new things in perf_mod have since been changed. May not be better, but not the same type of yuchiness. There are also a few additional new things, e.g. t_set_prefixf and t_unset_prefixf and their use in component_mod.F90. To be considered for the next merge I guess. Again, these or needed internally at the moment, but everything is fair game for removal for the public version?

File: cime_config/cesm/machines/config_batch.xml (hunk 1 above)
Comment (jgfouca) [inline cime_config/cesm/machines/config_batch.xml]: Done.

File: driver_cpl/shr/seq_flds_mod.F90 (hunk 10 above)
Comment (jgfouca) [inline driver_cpl/shr/seq_flds_mod.F90]: Comment added.

File: scripts/Testing/Testcases/HOMME_build.csh (hunk 3 above)
Comment (jgfouca) [inline scripts/Testing/Testcases/HOMME_build.csh]: Yes, I realize this is a hack. I was tasked with getting Homme testing into our nightly test suites and I thought the easiest way to do this was to make a HOMME testcase. I did not want to be in the situation where some of our tests run through create_test via test cases and others are unit-tests.

File: scripts/Tools/case.submit (hunk 4 above)
Comment (jgfouca) [inline scripts/Tools/case.submit]: Here's the log:

    BATCH SUBMISSION SO THAT 1ST BATCH HOLDS ON PASSED
    JOBID: -j 1234 or -jobid 1234
    TO THE <casename>.submit in Config folder

File: scripts/Tools/pointclm/pointCLM.py (hunk 5 above)
Comment (jgfouca) [inline scripts/Tools/pointclm/pointCLM.py]: You guys use pointCLM? I thought that was an ACME-created thing?

File: scripts/create_newcase (hunk 12 above)
Comment (jgfouca) [inline scripts/create_newcase]: @rljacob : this introduces an important problem. What am I supposed to do with features that were implemented poorly on the ACME side? Am I supposed to refactor the work to do it right? Or throw it out and ask the developers to reimplement once ACME is using ESMCI CIME?

Comment (rljacob) [inline scripts/create_newcase]: If we have time, refactor it.  Otherwise, ask the developer to reimplement with some very clear instructions.

Comment (jgfouca) [inline scripts/create_newcase]: OK, I am going to reject this change.

**Comment by jgfouca** [comment]
@jedwards4b , thanks for the close review. I have removed the cam_se stuff. Regarding the HOMME test, I may have to push back until there is better integration between create_test, update_acme_tests, the unit test framework, and our dashboard.

**Comment by jgfouca** [comment]
I think this PR is ready to go pending @mvertens concerns about PointCLM which I did not understand.

**Comment by rljacob** [comment]
Is PointCLM supposed to live somewhere else in the source?

**Comment by jgfouca** [comment]
@rljacob on the ACME side, it lives here: cime/scripts-acme/pointclm

File: scripts/Tools/pointclm/pointCLM.py (hunk 5 above)
Comment (mvertens) [inline scripts/Tools/pointclm/pointCLM.py]: the is no longer a pointclm (PTCLM) in the ncar esmci. Since it is a land-component specific utility it is in  components/clm/tools/shared/PTCLM (i.e. within the clm source).  It does not belong in esmci in my opinion.

[3:25] 
just a note from the CESM PTCLM:

[3:25] 
PTCLMmkdata is a python tool built on top of CLM tools and CESM scripts
for building datasets to run CLM "I" cases for data from Ameriflux Tower-sites,
or other user-supplied single-point datasets.

Comment (jgfouca) [inline scripts/Tools/pointclm/pointCLM.py]: @rljacob , can you advise me here? Is ACME's PointCLM filling the same roll as the PTCLM that existed in CESM?

Comment (rljacob) [inline scripts/Tools/pointclm/pointCLM.py]: I don't know.  But since it calls things in CIME, I can see how it wound up where it is.

Comment (mvertens) [inline scripts/Tools/pointclm/pointCLM.py]: I am pretty sure that ACME's PointCLM does fill the same roll as the PTCLM that existed in CESM.  You guys started effectively from CLM4.5 which also had the PTCLM capability. Not sure what changes ACME made. The key point is that this should be in the ALM directory not in the cime directory. That is what we did with PTCLM in our case.

Comment (jgfouca) [inline scripts/Tools/pointclm/pointCLM.py]: @mvertens , it is OK if we move after ACME adopts ESMCI?

Comment (ekluzek) [inline scripts/Tools/pointclm/pointCLM.py]: ​OK, there are several different flavors of single-point CLM. It sounds
like this is referring to the Python version of PTCLM that created CLM
datasets, and then setup a case for it. We decided that having a tool that
does both data-set creation AND interacts with cime/scripts to setup a case
-- was a bad idea.

So we split the functionality and we now have a Python script under:

components/clm/tools/shared/PTCLM

called PTCLMmkdata

that does the dataset creation part. It both creates the datasets as well
as creates user_nl_clm and shell_commands files that can be used in
conjunction with "create_newcase" in scripts. This is a much more
sustainable way to make the same functionality work. When it was hardwired
to "create_newcase" it seemed like it broke all the time. And the step of
creating data is a one-time step, while creating cases from data you've
setup is something you do a lot of. So it just made sense to separate the
two.

The ACME version of CLM and CIME is going to have both the PTCLM.py script
as well as PTCLMmkdata. Do you really want both? And do you really want to
have a script that is hardwired into it's use of "create_newcase"? We
decided that was a bad plan and went away from it -- does ACME really want
to retain that? Totally just questions for ACME to figure out. I'm just
giving the history and pointing out that you have both, and why we went to
the PTCLMmkdata method. And I am suggesting you might want to make the same
shift as well.

Thanks.

Feel free to ask me questions about all of this...

​

*Erik Kluzek, (CGD at NCAR)*

*National Center for Atmospheric Research*

*Boulder CO, *

*(off) (303)497-1326 (fax) (303)497-1348*

*(skype) ekluzek    (cell) (303)859-0183*

*Pronouns: he/his/him*

*------------------ Home page ------------------------*

*          http://www.cgd.ucar.edu/~erik <http://www.cgd.ucar.edu/~erik>*

*          https://staff.ucar.edu/users/erik
<https://staff.ucar.edu/users/erik>*

*!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*

File: scripts/Tools/case.submit (hunk 4 above)
Comment (jedwards4b) [inline scripts/Tools/case.submit]: I think that this was added because acme removed the requirement to use the case.submit script to start the job and so they are doing manually what that script is designed to automate.
